micro_http/src/server.rs

```diff
@@ -0,0 +1,551 @@
+    ///
+    /// This function is responsible for the data exchange with the clients and should
+    /// be called when we are either notified through `epoll` that we need to exchange
+    /// data with at least a client or when we don't need to perform any other operations
+    /// on this thread and we can afford to call it in a loop.
+    ///
+    /// Returns a collection of complete and valid requests to be processed by the user
+    /// of the server. Once processed, responses should be sent using `enqueue_responses`.
+    pub fn wait_server(&mut self) -> Result<Vec<(Token, Request)>> {
+        let mut parsed_requests: Vec<(Token, Request)> = vec![];
+        let mut events = vec![epoll::Event::new(epoll::Events::empty(), 0); MAX_CONNECTIONS];
+        let num_ev =
+            epoll::wait(self.epoll_fd, -1, &mut events[..]).map_err(|_| ServerError::IOError)?;
+        let mut streams_to_be_removed: Vec<Token> = vec![];
+        for e in events.iter().take(num_ev) {
+            match e.data as Token {
+                NEW_CONNECTION_EPOLL_TOKEN => {
+                    match self.handle_new_connection() {
+                        Err(ServerError::ServerFull) => continue,
+                        Err(error) => return Err(error),
+                        Ok(()) => {}
+                    };
+                }
+                token => {
+                    self.serve_connection(
+                        token,
+                        e.events,
+                        &mut streams_to_be_removed,
+                        &mut parsed_requests,
+                    )?;
+                }
+            }
+        }
+        for token in streams_to_be_removed {
+            self.del_stream(token);
+        }
+
+        Ok(parsed_requests)
+    }
+
+    /// Returns the file descriptor of the server's internal `epoll` structure.
+    pub fn get_epoll_fd(&self) -> RawFd {
+        self.epoll_fd
+    }
+
+    /// Enqueues the provided responses in the outgoing connection.
+    pub fn enqueue_responses(&mut self, responses: Vec<(Token, Response)>) -> Result<()> {
+        for (token, response) in responses {
+            if let Some(file_streamed_connection) = self.connections.get_mut(&token) {
+                file_streamed_connection
+                    .connection
+                    .enqueue_response(response);
+
+                epoll::ctl(
+                    self.epoll_fd,
+                    epoll::ControlOptions::EPOLL_CTL_MOD,
+                    file_streamed_connection.as_raw_fd(),
+                    epoll::Event::new(epoll::Events::EPOLLOUT, token as u64),
+                )
+                .map_err(|_| ServerError::IOError)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    // Accepts a new incoming connection and adds it to the `epoll` notification structure.
+    fn handle_new_connection(&mut self) -> Result<()> {
+        let (mut new_stream, _new_addr) = self.socket.accept().expect("Socket accept failed.");
```

Comment:
You can refactor this to be more rusty like in `virtio/vsock/unix/muxer.rs:366`.